### PR TITLE
DM-37720: use pybind wrappers, merge into one pybind shared lib

### PR DESCRIPTION
**Pull request by mwittgen**

### CODE REVIEW

File: python/lsst/jointcal/associations.cc [hunk 1]
```diff
@@ -36,46 +37,47 @@ namespace lsst {
...
+    });
 }
+}  // namespace
 
-PYBIND11_MODULE(associations, mod) {
-    py::module::import("lsst.jointcal.ccdImage");
-    py::module::import("lsst.sphgeom");
-    declareAssociations(mod);
+void wrapAssociations(lsst::cpputils::python::WrapperCollection &wrappers) {
+    wrappers.addInheritanceDependency("lsst.sphgeom");
```
Comment (erykoff) [inline python/lsst/jointcal/associations.cc:79]: This should be in the main `_jointcalLib.cc`.

File: python/lsst/jointcal/astrometryTransform.cc [hunk 2]
```diff
@@ -42,131 +37,159 @@ namespace lsst {
...
-                                                                                        "TanSipPixelToRaDec");
+void declareTanSipPixelToRaDec(lsst::cpputils::python::WrapperCollection &wrappers) {
+    using PyTanSipPixelToRaDec = py::class_<TanSipPixelToRaDec, std::shared_ptr<TanSipPixelToRaDec>, BaseTanWcs>;
+
+    wrappers.wrapType(PyTanSipPixelToRaDec(wrappers.module, "TanSipPixelToRaDec"), [](auto &mod, auto &cls) {});
 }
+}  // namespace
 
-PYBIND11_MODULE(astrometryTransform, mod) {
-    py::module::import("astshim.mapping");
```
Comment (erykoff) [inline python/lsst/jointcal/astrometryTransform.cc:150]: Does this need to be declared in `_jointcalLib.cc`?

**Review by erykoff — COMMENTED** [review]
A couple comments on imports.  But more importantly, I'm not sure why we need the compatibility stubs here?  Nobody is importing from jointcal, so isn't the key to fix any internal imports?

File: python/lsst/jointcal/associations.cc (hunk 1 above)
Comment (mwittgen) [reply python/lsst/jointcal/associations.cc:79]: Moved

File: python/lsst/jointcal/astrometryTransform.cc (hunk 2 above)
Comment (mwittgen) [reply python/lsst/jointcal/astrometryTransform.cc]: I don't see any dependencies on {{astshim.mapping}} here.

**Comment by mwittgen** [comment]
Rebased. Removed all compatibility stubs and updated tests accordingly.

File: python/lsst/jointcal/astrometryTransform.cc [hunk 3]
```diff
@@ -42,131 +37,159 @@ namespace lsst {
...
-                                                                                        "TanSipPixelToRaDec");
+void declareTanSipPixelToRaDec(lsst::cpputils::python::WrapperCollection &wrappers) {
+    using PyTanSipPixelToRaDec = py::class_<TanSipPixelToRaDec, std::shared_ptr<TanSipPixelToRaDec>, BaseTanWcs>;
+
+    wrappers.wrapType(PyTanSipPixelToRaDec(wrappers.module, "TanSipPixelToRaDec"), [](auto &mod, auto &cls) {});
 }
+}  // namespace
 
-PYBIND11_MODULE(astrometryTransform, mod) {
-    py::module::import("astshim");
```
Comment (erykoff) [inline python/lsst/jointcal/astrometryTransform.cc:150]: I thought this should be brought into the main file?

File: python/lsst/jointcal/_jointcalLib.cc [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    wrapPhotometryMappings(wrappers);
+    wrapPhotometryModels(wrappers);
+    wrapPhotometryTransform(wrappers);
+    wrapProjectionHandler(wrappers);
+    wrapStar(wrappers);
+    wrappers.finish();
+}
+
+}
+}
```
Comment (erykoff) [inline python/lsst/jointcal/_jointcalLib.cc:67]: This needs a newline at the end.

File: python/lsst/jointcal/astrometryTransform.cc (hunk 3 above)
Comment (mwittgen) [reply python/lsst/jointcal/astrometryTransform.cc:150]: Not a signature or inheritance of any classes used in these python bindings.
`astshim` is only used in `jointcal` python code. Having a closer look, there are unneeded
`#include "astshim.h"` in the C++ code that could be removed.

Comment (mwittgen) [reply python/lsst/jointcal/astrometryTransform.cc:150]: Ah, my mistake. Yes it's a signature dependency. Added.

File: python/lsst/jointcal/photometryTransform.cc [hunk 5]
```diff
@@ -8,7 +8,7 @@
  * See the COPYRIGHT file at the top-level directory of this distribution
  * for details of code ownership.
  *
- * This program is free software: you can redistribute it and/or modify
+ * This program is free software: you can redistribute it and/or wrappersify
```
Comment (erykoff) [inline python/lsst/jointcal/photometryTransform.cc:11]: Global replace fail! 😂

File: python/lsst/jointcal/projectionHandler.cc [hunk 6]
```diff
@@ -8,7 +8,7 @@
  * See the COPYRIGHT file at the top-level directory of this distribution
  * for details of code ownership.
  *
- * This program is free software: you can redistribute it and/or modify
+ * This program is free software: you can redistribute it and/or wrappersify
```
Comment (erykoff) [inline python/lsst/jointcal/projectionHandler.cc:11]: Global replace fail.

File: python/lsst/jointcal/star.cc [hunk 7]
```diff
@@ -8,7 +8,7 @@
  * See the COPYRIGHT file at the top-level directory of this distribution
  * for details of code ownership.
  *
- * This program is free software: you can redistribute it and/or modify
+ * This program is free software: you can redistribute it and/or wrappersify
```
Comment (erykoff) [inline python/lsst/jointcal/star.cc:11]: Global replace fail.

**Review by erykoff — APPROVED** [review]
Looks good except for a few funny inadvertent license replacements.

**Pull request merged**